<a href="https://colab.research.google.com/github/Octa-p/DWDM21/blob/main/MiniExam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### ข้อสอบ

จาก data ```flights.csv``` และ ```airports.csv``` ใน https://drive.google.com/drive/folders/1EFo1_uQgWkRiXjezL0T4YwteYrsqU5MV?usp=sharing
0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด
1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด
2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์
3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด
4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด
5. วันธรรมดา หรือวันเสาร์อาทิตย์มีเครื่องบินขาเข้ามากกว่ากัน (วันจันทร์ = 1, วันอังคาร = 2, วันพุธ = 3,...)

ให้นักศึกษาทำข้อสอบ 2 ข้อจากที่แสดงข้างบนโดยกรอกรหัสนักศึกษาลงใน function ```gen_exams()``` แล้ว run เพื่อให้ได้ข้อสอบที่นักศึกษาต้องทำ 2 ข้อ (ข้อละ 3 คะแนน)
และให้นักศึกษาคิดโจทย์เองอีก 1 ข้อ (4 คะแนน)


In [1]:
import pandas as pd

In [2]:
from google.colab import drive 
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os

In [4]:
path = '/content/drive/My Drive/DWDM2021_DATA'

In [5]:
flights = pd.read_csv(os.path.join(path,'flights.csv'))
flights

วันในเดือน  วันในสัปดาห์  ... ความล่าช้าขาออก  ความล่าช้าขาเข้า
0                19             5  ...              -3                 1
1                19             5  ...               0                -8
2                19             5  ...              -4               -15
3                19             5  ...              28                24
4                19             5  ...              -6               -11
...             ...           ...  ...             ...               ...
2702213          19             6  ...               1                -3
2702214          19             6  ...              -2                -3
2702215          19             6  ...              18                 9
2702216          19             6  ...              10                 7
2702217          19             6  ...               0               -13

[2702218 rows x 7 columns]

In [6]:
airports = pd.read_csv(os.path.join(path,'airports.csv'))
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [7]:
def gen_exams(ID):
    import math
    a = int(ID[-1])%6
    b = int(ID[-1])%3
    print(f'รหัสนักศึกษา {ID} ทำข้อ {a} และ {b if a != b else 5-b}')

ตัวอย่างการ run

In [8]:
gen_exams('623020042-0')

รหัสนักศึกษา 623020042-0 ทำข้อ 0 และ 5


### **หาค่า Missing ในตารางก่อน**

In [9]:
flights.isnull()

วันในเดือน  วันในสัปดาห์  ...  ความล่าช้าขาออก  ความล่าช้าขาเข้า
0             False         False  ...            False             False
1             False         False  ...            False             False
2             False         False  ...            False             False
3             False         False  ...            False             False
4             False         False  ...            False             False
...             ...           ...  ...              ...               ...
2702213       False         False  ...            False             False
2702214       False         False  ...            False             False
2702215       False         False  ...            False             False
2702216       False         False  ...            False             False
2702217       False         False  ...            False             False

[2702218 rows x 7 columns]

In [10]:
flights.isnull().any()

วันในเดือน            False
วันในสัปดาห์          False
สายการบิน             False
รหัสสนามบินต้นทาง     False
รหัสสนามบินปลายทาง    False
ความล่าช้าขาออก       False
ความล่าช้าขาเข้า      False
dtype: bool

In [11]:
airports.isnull()

รหัสสนามบิน  เมือง    รัฐ  ชื่อสนามบิน
0          False  False  False        False
1          False  False  False        False
2          False  False  False        False
3          False  False  False        False
4          False  False  False        False
..           ...    ...    ...          ...
360        False  False  False        False
361        False  False  False        False
362        False  False  False        False
363        False  False  False        False
364        False  False  False        False

[365 rows x 4 columns]

In [12]:
airports.isnull().any()

รหัสสนามบิน    False
เมือง          False
รัฐ            False
ชื่อสนามบิน    False
dtype: bool

**ทั้ง 2 ตารางที่ให้มาไม่มีค่า Missing เลย เพราะเป็น False หมดทุกช่อง**

### **รวมตารางใหม่ เพื่อให้ดูง่ายขึ้น**

In [13]:
arrive_plane = flights.merge(airports,how='left',left_on='รหัสสนามบินปลายทาง',right_on='รหัสสนามบิน')
arrive_plane #how คือเอาตารางไหนเป็นตารางหลัก จะเห็นได้ว่ารหัสสนามบินปลายทางและรหัสสนามบิน มีรหัสที่ตรงกันเพราะการเชื่อมตาราง

วันในเดือน  ...                                ชื่อสนามบิน
0                19  ...                        Miami International
1                19  ...              John F. Kennedy International
2                19  ...               Salt Lake City International
3                19  ...                 Detroit Metro Wayne County
4                19  ...                  Los Angeles International
...             ...  ...                                        ...
2702213          19  ...               Salt Lake City International
2702214          19  ...  Louis Armstrong New Orleans International
2702215          19  ...   Hartsfield-Jackson Atlanta International
2702216          19  ...                      Memphis International
2702217          19  ...          Minneapolis-St Paul International

[2702218 rows x 11 columns]

**ตัด Column รหัสสนามบินต้นทางออก เพื่อให้ดูง่ายขึ้นว่าตารางนี้คือตารางเครื่องบินขาเข้า**

In [14]:
arrive_plane.drop(['รหัสสนามบินต้นทาง'], axis=1, inplace=True)

**ตารางเครื่องบินขาเข้า**

In [15]:
arrive_plane

วันในเดือน  ...                                ชื่อสนามบิน
0                19  ...                        Miami International
1                19  ...              John F. Kennedy International
2                19  ...               Salt Lake City International
3                19  ...                 Detroit Metro Wayne County
4                19  ...                  Los Angeles International
...             ...  ...                                        ...
2702213          19  ...               Salt Lake City International
2702214          19  ...  Louis Armstrong New Orleans International
2702215          19  ...   Hartsfield-Jackson Atlanta International
2702216          19  ...                      Memphis International
2702217          19  ...          Minneapolis-St Paul International

[2702218 rows x 10 columns]

**เช็คค่า Missing**

In [16]:
arrive_plane.isnull().any()

วันในเดือน            False
วันในสัปดาห์          False
สายการบิน             False
รหัสสนามบินปลายทาง    False
ความล่าช้าขาออก       False
ความล่าช้าขาเข้า      False
รหัสสนามบิน           False
เมือง                 False
รัฐ                   False
ชื่อสนามบิน           False
dtype: bool

**ทุกคอลัมป์เป็น False แสดงว่าไม่มีค่า Missing**

In [17]:
outbound_plane = flights.merge(airports,how='left',left_on='รหัสสนามบินต้นทาง',right_on='รหัสสนามบิน')
outbound_plane #how คือเอาตารางไหนเป็นตารางหลัก จะเห็นได้ว่ารหัสสนามบินต้นทางและรหัสสนามบิน มีรหัสที่ตรงกันเพราะการเชื่อมตาราง

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 11 columns]

**ตัด Column รหัสสนามบินปลายทางออก เพื่อให้ดูง่ายขึ้นว่าตารางนี้คือตารางเครื่องบินขาออก**

In [18]:
outbound_plane.drop(['รหัสสนามบินปลายทาง'], axis=1, inplace=True)

**ตารางเครื่องบินขาออก**

In [19]:
outbound_plane

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 10 columns]

**เช็คค่า Missing**

In [20]:
outbound_plane.isnull().any()

วันในเดือน           False
วันในสัปดาห์         False
สายการบิน            False
รหัสสนามบินต้นทาง    False
ความล่าช้าขาออก      False
ความล่าช้าขาเข้า     False
รหัสสนามบิน          False
เมือง                False
รัฐ                  False
ชื่อสนามบิน          False
dtype: bool

**ทุกคอลัมป์เป็น False แสดงว่าไม่มีค่า Missing**

### **ข้อ 0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด**

**รัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุด**

In [21]:
MaxArrive = arrive_plane.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False) #ascending=False คือเรียงจากมากไปน้อย
MaxArrive.iloc[:1,[3]]

รหัสสนามบินปลายทาง
รัฐ                    
CA               380429

**รัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดคือรัฐ CA มีจำนวนเครื่องบินขาเข้าทั้งหมด 380,429 ลำ**

**รัฐที่มีจำนวนเครื่องบินขาเข้าน้อยที่สุด**

In [22]:
MinArrive = arrive_plane.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=True) #ascending=False คือเรียงจากน้อยไปมาก
MinArrive.iloc[:1,[3]]

รหัสสนามบินปลายทาง
รัฐ                    
AK                 7711

**รัฐที่มีจำนวนเครื่องบินขาเข้าน้อยที่สุดคือรัฐ AK มีจำนวนเครื่องบินขาเข้าทั้งหมด 7,711 ลำ**

### **ข้อ 5. วันธรรมดาหรือวันเสาร์-อาทิตย์มีเครื่องบินขาเข้ามากกว่ากัน (วันจันทร์ = 1, วันอังคาร = 2, วันพุธ = 3,...)**

In [23]:
MaxDay = arrive_plane.groupby('วันในสัปดาห์').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False) #ascending=False คือเรียงจากมากไปน้อย
MaxDay.iloc[:,[2]]

รหัสสนามบินปลายทาง
วันในสัปดาห์                    
1                         407837
4                         406563
3                         403072
2                         397594
5                         396387
7                         372228
6                         318537

**จากตาราง MaxDay จะพบว่า**

**วันในสัปดาห์ (1) คือ วันจันทร์ มีเครื่องบินขาเข้าทั้งหมด 407,837 ลำ**

**วันในสัปดาห์ (4) คือ วันพฤหัสบดี มีเครื่องบินขาเข้าทั้งหมด 406,563 ลำ**

**วันในสัปดาห์ (3) คือ วันพุธ มีเครื่องบินขาเข้าทั้งหมด 403,072 ลำ**

**วันในสัปดาห์ (2) คือ วันอังคาร มีเครื่องบินขาเข้าทั้งหมด 397,594 ลำ**

**วันในสัปดาห์ (5) คือ วันศุกร์ มีเครื่องบินขาเข้าทั้งหมด 396,387 ลำ**

**วันในสัปดาห์ (7) คือ วันอาทิตย์ มีเครื่องบินขาเข้าทั้งหมด 372,228 ลำ**

**วันในสัปดาห์ (6) คือ วันเสาร์ มีเครื่องบินขาเข้าทั้งหมด 318,537 ลำ**

**จะได้เป็นวันธรรมดามีเครื่องบินขาเข้าทั้งหมด 2,011,453 ลำ**

**และวันเสาร์-อาทิตย์มีเครื่องบินขาเข้าทั้งหมด 690,765 ลำ**

**ดังนั้นวันธรรมดาจึงมีจำนวนเครื่องบินขาเข้ามากกว่าวันเสาร์-อาทิตย์**

### **(ข้อที่คิดเอง) หาวันในสัปดาห์เฉลี่ยของสายการบิน DL**

In [48]:
Date = arrive_plane['สายการบิน'] == 'DL'
Date_mean = arrive_plane[Date]['วันในสัปดาห์'].mean()
Date_mean

3.9203447074205373

**สายการบิน DL ออกเที่ยวบินเฉลี่ยเป็นวันที่ 3 ในสัปดาห์ หรือวันพุธ**